cpp/src/gandiva/hash_utils.cc

```diff
@@ -0,0 +1,137 @@
+    // The buffer size is the hash size + null character
+    int sha256_result_length = 65;
+    return HashUtils::GetHash(context, message, message_length, EVP_sha256(),
+							  sha256_result_length, out_length);
+  }
+  const char* HashUtils::HashUsingSha128(int64_t context,
+                                         const void* message,
+                                         size_t message_length,
+                                         int32_t *out_length) {
+    // The buffer size is the hash size + null character
+    int sha128_result_length = 41;
+    return HashUtils::GetHash(context, message, message_length, EVP_sha1(),
+							  sha128_result_length, out_length);
+  }
+
+  const char* HashUtils::GetHash(int64_t context,
+                                 const void* message,
+                                 size_t message_length,
+                                 const EVP_MD *hash_type,
+                                 int result_buf_size,
+                                 int32_t *out_length) {
+    EVP_MD_CTX *md_ctx = EVP_MD_CTX_new();
+
+    if (md_ctx == nullptr) {
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
+									   "for SHA processing.");
+      return "";
+    }
+
+    int evp_success_status = 1;
+
+    if (EVP_DigestInit_ex(md_ctx, hash_type, nullptr) != evp_success_status) {
+      HashUtils::ErrorMessage(context, "Could not obtain the hash "
+									   "for the defined value.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    if (EVP_DigestUpdate(md_ctx, message, message_length) != evp_success_status) {
+      HashUtils::ErrorMessage(context, "Could not obtain the hash for "
+									   "the defined value.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    int hash_size = EVP_MD_size(hash_type);
+    auto* result = static_cast<unsigned char*>(OPENSSL_malloc(hash_size));
+
+    if (result == nullptr) {
+      HashUtils::ErrorMessage(context, "Could not allocate memory "
+									   "for SHA processing.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    unsigned int result_length;
+    EVP_DigestFinal_ex(md_ctx, result, &result_length);
+
+    int tmp_buf_len = 4;
+
+    auto hex_buffer =
+        reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, tmp_buf_len));
+
+    auto result_buffer =
+        reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, result_buf_size));
+
+    CleanCharArray(result_buffer);
+    CleanCharArray(hex_buffer);
+
+    if (hex_buffer == nullptr || result_buffer == nullptr) {
+      gdv_fn_context_set_error_msg(context, "Could not allocate memory "
+                                       "for the result buffers.");
+      EVP_MD_CTX_free(md_ctx);
+      return "";
+    }
+
+    for (unsigned int j = 0; j < result_length; j++) {
+      unsigned char hex_number = result[j];
+      snprintf(hex_buffer, tmp_buf_len, "%02x", hex_number);
```

Comment:
it seems like you should snprintf to the final result_buffer directly and update the pointer using the returned value? (you porbably also want to at least DCHECK the result.

Reply:
Suggestion applied